File: demo/backend/typescript/src/procedures/list-users.ts
import { t } from "@canmi/seam-server";
import type { ProcedureDef } from "@canmi/seam-server";

interface UserSummary {
  id: number;
  name: string;
}

const USERS: UserSummary[] = [
  { id: 1, name: "Alice" },
  { id: 2, name: "Bob" },
  { id: 3, name: "Charlie" },
];

export const listUsers: ProcedureDef<Record<string, never>, UserSummary[]> = {
  input: t.object({}),
  output: t.array(t.object({ id: t.uint32(), name: t.string() })),
  handler: () => {
    return USERS;
  },
};

File: examples/standalone/server-rust/src/procedures/list_users.rs
/* examples/standalone/server-rust/src/procedures/list_users.rs */

use seam_server::{SeamError, SeamType, seam_procedure};
use serde::{Deserialize, Serialize};

#[derive(Deserialize, SeamType)]
pub struct ListUsersInput {}

#[derive(Serialize, SeamType)]
pub struct UserSummary {
  pub id: u32,
  pub name: String,
}

#[seam_procedure(name = "listUsers")]
pub async fn list_users(_input: ListUsersInput) -> Result<Vec<UserSummary>, SeamError> {
  Ok(vec![
    UserSummary { id: 1, name: "Alice".to_string() },
    UserSummary { id: 2, name: "Bob".to_string() },
    UserSummary { id: 3, name: "Charlie".to_string() },
  ])
}
